### ISSUE TITLE
add documentation of the meaning of resumeOffset=0 in twisted.words.protocols.irc

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7775]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#7775|
|Type|enhancement|
|Created|2015-02-07 04:19:49Z|
|Branch|https://github.com/twisted/twisted/tree/irc-dcc-resume-7775|

See [[[#6494](https://github.com/twisted/twisted/issues/6494)](https://github.com/twisted/twisted/issues/6494)#comment:25 here].

Eeshan, this is a follow-on from #6494 so if you could address it that would be great.

Attachments:

* [bugfix_7775.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc0/bc0e8bc45a5e26586a4ab517d0a00e4a1228a9f3/9bb5fbb22685f6f47784c1f490854c85cdeb67b5.patch) (6592 bytes) - added by eeshangarg on 2015-04-15 21:23:14Z - 
* [bugfix_7775_2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc0/bc0e8bc45a5e26586a4ab517d0a00e4a1228a9f3/1ed3811bf4bfb67fb000ebc6995f2890b6b03cfd.patch) (7919 bytes) - added by eeshangarg on 2015-04-20 14:37:11Z - 
* [bugfix_7775_3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bc0/bc0e8bc45a5e26586a4ab517d0a00e4a1228a9f3/440d692d276c5b6c278e20315ff552dc53bcd615.patch) (7621 bytes) - added by eeshangarg on 2015-04-20 15:42:01Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7775 7775
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_irc_dcc_resume_7775 branches/irc-dcc-resume-7775
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1423282789972348 1423282789972348
changetime__1429866104309730 1429866104309730
version__None None
owner__eeshangarg eeshangarg

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_1"></a>|@eeshangarg commented|
|-|-|

Hello, Mr. glyph!

Thanks a lot for asking me to work on this. It really is an honor. I would love to work on this! :-) I have a couple of questions that I would like to ask.

Admittedly, much of what I am going to write in the next paragraph has already been clearly expressed by Mr. adiroiban in his comments in [ticket:6494]. However, before I start working on this, I think it would be a great learning experience for a beginner like me if you allow me to give a short description of what I understand about the problem at hand as this will allow you to point out if I am wrong about something or everything (a more than likely possibility!).

So, the docstring for twisted.words.protocols.irc.DccFileReceive already has a one line description of what resumeOffset is supposed to do, but I am guessing that it is not what it actually does, hence this ticket. So DccFileReceive calls the constructor for it's parent class DccFileReceiveBasic which assigns a Boolean value to it's (DccFileReceiveBasic's) attribute self.resume based on whether resumeOffset contains a zero or non-zero value, which, if True, then the connectionMade method for DccFileReceive opens the file in append mode and writes the data to it; saying that it is starting from %d bytes where %d is the number of bytes as returned by file.tell(). Also, as you can see, in DccFileReceiveBasic, before self.resume is assigned a Boolean, self.bytesReceived is assigned the numerical value of resumeOffset and according to me, this is the only place where the numeric value for resumeOffset seems to be doing something. What I don't understand is, what role does self.bytesReceived play in the whole thing? Does it influence in any way the position in the file from where the data transfer will resume when the connectionMade method in DccFileReceive is called in the case that self.resume is True?

One more thing that I'm not entirely sure about, will this ticket involve changing the code in some way or the other, if so, to what extent?

Again, with your guidance, I would love to work on this. Thanks a lot for letting me be a part of this amazing community! Also, I am sorry if I am being too verbose here and if my description is disorderly (it probably is), I'm trying my best to get better at this.

Warm regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

> what role does self.bytesReceived play in the whole thing? Does it influence in any way the position in the file from where the data transfer will resume when the connectionMade method in DccFileReceive is called in the case that self.resume is True?

I'm not sure what `bytesReceived` plays - you'll have to do a bit more digging on that front :).

> One more thing that I'm not entirely sure about, will this ticket involve changing the code in some way or the other, if so, to what extent?

This ticket should not involve changing the code; simply document what `resumeOffset` really means; it just says "documentation" :).  If in the course of doing so, you discover that it's not possible to sensibly document, then close it and open a new one that fixes its behavior.

> Again, with your guidance, I would love to work on this. Thanks a lot for letting me be a part of this amazing community! Also, I am sorry if I am being too verbose here and if my description is disorderly (it probably is), I'm trying my best to get better at this.

The community is really nothing without its participants, so thank _you_ for participating in it :-).

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_4"></a>|@eeshangarg removed owner|
|-|-|

Hello!

I just submitted a patch for this (bugfix_7775.patch​). I worked on this at the PyCon sprints and found that the resumeOffset argument does not work as it is documented.

I talked to Glyph about this and he said:

"""
It's a bug. It's a gross violation of specification.
"""

I also removed this code snippet/comment <https://github.com/twisted/twisted/blob/trunk/twisted/words/protocols/irc.py#L2346>, as it uses methods that no longer exist.

If this needs further work, I would love to work on it! :-)

Thanks to Glyph and lvh for their support, encouragement and guidance. It is a huge honor for me to learn from you. :-)

Warm regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban set owner to @eeshangarg|
|-|-|

Hi,

Many thanks for the changes. It looks really great!

For readying and writing files you can try to use FilePath... with setContent and getContent

```
file = FilePath(self.mktemp())
file.setContent('bla')
result = file.getContent()
```


-----

This code is duplicated in multiple places
```
protocol = irc.DccFileReceive(filename) 
transport = StringTransport() 
protocol.makeConnection(transport) 
```

Maybe you can refactor it into a factory helper like this

```

def makeConnectedDccFileReceive(self, resumeOffset=None)
  protocol = irc.DccFileReceive(filename, resumeOffset=resumeOffset) 
  transport = StringTransport() 
  protocol.makeConnection(transport) 
  return protocol

# And use it like
protocol = self.makeConnectedDccFileReceive()

```

In a similar way:

```
def allDataReceivedForProtocol(self, protocol, data):
  protocol.dataReceived(data) 
  protocol.connectionLost(None)
 
self.allDataReceivedForProtocol(protocol, 'bla bla')

```

Also, not sure what to say about the indentation... but this also a minor comment.

The above comments are all minor but I hope that will help improve the readability of the tests.
Since we don't have tests for the test code itself it is very important that tests are well documented and easy to follow and understand.

Please let me know what do you thing about the above comments.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_6"></a>|@eeshangarg removed owner|
|-|-|

Thanks for the review, adiroiban!

I just added a new patch (bugfix_7775_2.patch) with all of the changes that you talked about. :-)

Let me know if this needs further work! :-)

Regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban commented|
|-|-|

(In [44511]) Branching to irc-dcc-resume-7775.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_8"></a>|@adiroiban set owner to @eeshangarg|
|-|-|

Thanks. Changes looks good.

I think that self.resumeOffset should be private as external parties should not core about it.

Also, for logging message I think that is better to use self.file.tell() to check the position of the cursor as visible to the OS.


-------

Side comment:

Documentation for resumeOffset is a bit hidden in DccFileReceiveBasic but I think is ok.

I have no idea why do we need DccFileReceiveBasic as it is only used in DccFileReceive and it adds the self.resume public member on top of resumeOffset :(

Unfortunately it is a public API and we need to spend extra effort to deprecated it... so it should be ok.

------


Please check the branch attached to this ticket and let me know if you find any problem in making resumeOffset private and using file.tell()

Builders triggered.


Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_9"></a>|@eeshangarg removed owner|
|-|-|

Hi, adiroiban!

I totally agree with the changes you proposed. It makes sense to make self.resumeOffset non-public, as I personally cannot come up with a solid use case right now that hints at a future or present possibility of an external party explicitly caring about the value of the self.resumeOffset attribute. If we do find such a use case, then I think it would be much easier to make a non-public attribute public (if needed) than doing the opposite. :-)

Also, now that I think about it, it would be better to use file.tell() in the relevant log message than simply logging the value of resumeOffset.

So, +1 for both of the changes that you put forth.

Thanks for the review!

Regards,
Eeshan Garg

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

Sorry for the first review as I done it in hurry.
I did a new review.

We need to be more explicit about what data types we use to simplify the future porting to Python3. FilePath.setContent should receive bytes and we should explicitly mark byte strings.

For test_cannotOpenFile and test_setOverwrite ... the docstring was describing 2 cases but the test was testing only one.

The docstring should describe the specific case which is actually tested.

I have create a new test for the explicit case when file can not be opened and instead of OSError it gets IOError.

Please check latest changes from this branch.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="eeshangarg's avatar" src="https://avatars.githubusercontent.com/u/7251823?s=50" width="50" height="50">](https://github.com/eeshangarg)<a name="note_11"></a>|@eeshangarg commented|
|-|-|

Hey, adiroiban!

I took a look at the latest changes, and it's a +1 from me. If you feel there is anything else that I could improve or add, please let me know!. :-)

Thanks a lot for all of the reviews! I learned a lot from them! :-)

Regards,
Eeshan

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_12"></a>|@adiroiban set owner to @eeshangarg|
|-|-|

My test was not ok :( Your initial test was ok :)
I have updated it now.
Sorry for the trouble.

Sending this to buildbot and will merge if all is ok and meanwhile I will not find other problems.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set status to `closed`|
|-|-|

(In [44548]) Merge irc-dcc-resume-7775: Fix resumeOffset in t.w.protocol.irc.DccFileReceive.

Author: eeshangarg
Reviewers: adiroiban
Fixes: [#7775](https://github.com/twisted/twisted/issues/7775)
